### PR TITLE
o/configstate: don't pass --root=/ when masking/unmasking/enabling/disabling services

### PR DESCRIPTION
**Pull request by stolowski**
Based on https://github.com/snapcore/snapd/pull/10003 ; this branch pulls the new config option from 10003 as it's the only one that can be easily tested.

This fixes the problem experienced with #10003 where enabling core service back after a reboot would fail with "Unit .. is masked" error (despite of calling systemctl unmask...).

There seems to be a weird asymmetry in systemd behavior because in configcore handling code we pass --root=/ everywhere around systemctl calls for core services, yet it fails only when enabling something back (also it may be specific to core because AFAIR all combinations worked when I tested such scenarios manually on 20.04 classic).

I think the right way is not to pass --root=/ as (AFAIU) this bypasses systemd, which is something we probably should avoid (I suppose this may mean we will now fail on corner cases where we wouldn't notice previously). This is also in line with how we call systemctl in wrappers/services.go. Note this doesn't affect emulation mode.

### CODE REVIEW

File: tests/nested/core/coreconfig-services/task.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+execute: |
+  #shellcheck source=tests/lib/nested.sh
+  . "$TESTSLIB/nested.sh"
+  nested_exec "systemctl status systemd-resolved.service" | MATCH "Active: +active"
+
+  echo "Disabling systemd-resolved service"
+  nested_exec "sudo snap set system service.systemd-resolved.disable=true"
+  nested_exec "systemctl status systemd-resolved.service" | MATCH "Active: +inactive"
+  nested_exec "sudo reboot"
```
Comment (pedronis) [inline tests/nested/core/coreconfig-services/task.yaml:13]: is this reboot right?  without the waiting?

File: overlord/configstate/configcore/services.go [hunk 2]
```diff
@@ -144,7 +145,7 @@ func switchDisableService(serviceName string, disabled bool, opts *fsOnlyContext
 	if opts != nil {
 		sysd = systemd.NewEmulationMode(opts.RootDir)
 	} else {
-		sysd = systemd.NewUnderRoot(dirs.GlobalRootDir, systemd.SystemMode, &sysdLogger{})
```
Comment (pedronis) [inline overlord/configstate/configcore/services.go:147]: there are more places in  configcore that use NewUnderRoot should we change all of them?

**Review by pedronis — COMMENTED** [review]
change looks good, but question about the other similar code we have and about the test

File: tests/nested/core/coreconfig-services/task.yaml (hunk 1 above)
Comment (stolowski) [reply tests/nested/core/coreconfig-services/task.yaml:13]: ah, nope, copy-paste error, thanks!

File: overlord/configstate/configcore/services.go (hunk 2 above)
Comment (stolowski) [reply overlord/configstate/configcore/services.go:147]: Almost certainly yes. As discussed on the channel it will be a followup (having them spread-tested will required some effort and might be tricky).

**Review by pedronis — APPROVED** [review]
thank you

**Review by mvo5 — APPROVED** [review]
Thanks for getting to the bottom of this

File: tests/nested/core/coreconfig-services/task.yaml [hunk 3]
```diff
@@ -0,0 +1,21 @@
+summary: Disable and enable back core services via snap set with reboot.
+
+systems: [ubuntu-18.04-64]
```
Comment (anonymouse64) [inline tests/nested/core/coreconfig-services/task.yaml:3]: couldn't you just make this

```suggestion
systems: [ubuntu-18.04-64, ubuntu-20.04-64]
```

and then this test will run for uc20 and uc18? then we don't need to duplicate the same file. I guess I thought that tests/nested/core20 was for things that were uc20 specific, but if a test doesn't need to be uc20 specific then it can be run on uc20 here too (though certainly correct me if I'm wrong)

also why don't we run this with uc16 in the nested environment?

**Review by anonymouse64 — COMMENTED** [review]
lgtm, but question about the spread tests

File: tests/nested/core/coreconfig-services/task.yaml (hunk 3 above)
Comment (mvo5) [reply tests/nested/core/coreconfig-services/task.yaml:3]: IIRC we do not use systemd-resolved in UC16, that was still resolveconf back in the day. But I agree about resuing the file if that is possible

Comment (stolowski) [reply tests/nested/core/coreconfig-services/task.yaml:3]: For some reason in spread.yaml we have separate suites, and for tests/nested/core we have:
`systems: [ubuntu-16.04-64, ubuntu-18.04-64]` so this wouldn't work without changing spread.yaml.

From the quick scan the setup for tests/nested/core and tests/nested/core20 looks similar (if not identical) but maybe there are deeper reasons for the split; CC @sergiocazzolato

Comment (mvo5) [reply tests/nested/core/coreconfig-services/task.yaml:3]: Thanks for looking into this! We can always do that in a followup, it's fine for now as long as we have it on the radar.

Comment (anonymouse64) [reply tests/nested/core/coreconfig-services/task.yaml:3]: ah thanks, I forgot that systemd-resolved is not a thing in uc16... 
also regarding not duplicating these test files, I think that's fine to do in a followup if it requires changes to the spread.yaml, etc.

Comment (sergiocazzolato) [reply tests/nested/core/coreconfig-services/task.yaml:3]: @stolowski @anonymouse64 
The main differences between core and core20 suites are:
 . default systems to run the suite
 . core20 uses NESTED_ENABLE_TPM=true and NESTED_ENABLE_SECURE_BOOT=true
So if you run a test for uc20 in core suite, it won't use neither secure boot nor tpm.

**Comment by mvo5** [comment]
If someone has cargo-culted a gadget default like
```
defaults:
   system:
     service:
       rsyslog:
         disable: true
```
on core18 this will break them  because there is no rsyslog installed on core18. The old code did just do the filesystem masking/unmasking so it did not matter. But the new code is now actually talking to systemd and that knows that there is no rsyslog and fails.

So I move this to 2.50 for now and we need to first inspect all gadgets to ensure nothing was cargo-culted like this.

**Comment by stolowski** [comment]
> So I move this to 2.50 for now and we need to first inspect all gadgets to ensure nothing was cargo-culted like this.

Another solution (discussed with @pedronis on irc) is to check whether service exists before doing anything about it, I'll explore this idea.

**Comment by stolowski** [comment]
> > So I move this to 2.50 for now and we need to first inspect all gadgets to ensure nothing was cargo-culted like this.
> 
> Another solution (discussed with @pedronis on irc) is to check whether service exists before doing anything about it, I'll explore this idea.

This will be possible with https://github.com/snapcore/snapd/pull/10083 (after which this PR will be updated to check if service is installed).

**Comment by stolowski** [comment]
Updated to use Status().

**Review by mvo5 — APPROVED** [review]
Very nice, thanks for this update!

**Review by pedronis — APPROVED** [review]
the new code looks right to me

**Comment by mvo5** [comment]
I merged this now, the nested 18.04 failure is unrelated and also visible in other PRs like #10072 - I also run the branch manually:
```
$ spread -debug -v qemu-nested:ubuntu-18.04-64:tests/nested/manual/snapd-refresh-from-old
...
2021-04-09 13:58:42 Preparing qemu-nested:ubuntu-18.04-64:tests/nested/manual/ (qemu-nested:ubuntu-18.04-64)...
2021-04-09 13:59:37 Preparing qemu-nested:ubuntu-18.04-64:tests/nested/manual/snapd-refresh-from-old:edge_first (qemu-nested:ubuntu-18.04-64)...
2021-04-09 14:04:27 Executing qemu-nested:ubuntu-18.04-64:tests/nested/manual/snapd-refresh-from-old:edge_first (qemu-nested:ubuntu-18.04-64) (1/2)...
2021-04-09 14:05:41 Restoring qemu-nested:ubuntu-18.04-64:tests/nested/manual/snapd-refresh-from-old:edge_first (qemu-nested:ubuntu-18.04-64)...
2021-04-09 14:05:42 Preparing qemu-nested:ubuntu-18.04-64:tests/nested/manual/snapd-refresh-from-old:latest_only (qemu-nested:ubuntu-18.04-64)...
2021-04-09 14:07:53 Executing qemu-nested:ubuntu-18.04-64:tests/nested/manual/snapd-refresh-from-old:latest_only (qemu-nested:ubuntu-18.04-64) (2/2)...
2021-04-09 14:08:03 Restoring qemu-nested:ubuntu-18.04-64:tests/nested/manual/snapd-refresh-from-old:latest_only (qemu-nested:ubuntu-18.04-64)...
2021-04-09 14:08:04 Restoring qemu-nested:ubuntu-18.04-64:tests/nested/manual/ (qemu-nested:ubuntu-18.04-64)...
2021-04-09 14:08:07 Restoring qemu-nested:ubuntu-18.04-64 (qemu-nested:ubuntu-18.04-64)...
2021-04-09 14:08:07 Discarding qemu-nested:ubuntu-18.04-64...
2021-04-09 14:08:07 Successful tasks: 2
2021-04-09 14:08:07 Aborted tasks: 0
```
and it'f fine here.
